thunder/executors/cudagraphex.py

```diff
@@ -0,0 +1,246 @@
+    outputs: list[Proxy],
+) -> Callable:
+    from inspect import Parameter, Signature
+
+    region_fn_params = (
+        Parameter(getattr(param, "name", f"arg{i}"), Parameter.POSITIONAL_ONLY) for i, param in enumerate(inputs)
+    )
+
+    region_fn_signature = Signature(region_fn_params)
+
+    def region_fn():
+        pass
+
+    region_fn.__signature__ = region_fn_signature
+    region_fn.__name__ = fn_name
+
+    region_trace = TraceCtx(region_fn)
+    region_trace.bound_symbols = bsyms
+    region_trace.args = inputs
+    region_trace.kwargs = {}
+    region_trace.bound_symbols.append(prims.python_return.bind(outputs, output=()))
+
+    return region_trace.python_callable()
+
+
+class CUDAGraphExecutor(FusionExecutor):
+    def __init__(self, name: Hashable):
+        super().__init__(name, version=torch.version.cuda)
+
+    def fuse(self, region: Region, fusion_counter: int, num_static_inputs: None | int = None) -> BoundSymbol:
+        inputs = [unvariableify(inp) for inp in sorted(region.inputs, key=lambda var: var.proxy.name)]
+        outputs = [unvariableify(out) for out in sorted(region.outputs, key=lambda var: var.proxy.name)]
+
+        fusion_name = f"CUDAGraph{fusion_counter}"
+        fusion_callable: Callable = make_callable(f"{fusion_name}_fn", region.bound_symbols, inputs, outputs)
+        fusion_callable = CUDAGraphCallable(fusion_callable, num_static_inputs)
+
+        fusion_sym = Symbol(fusion_name, meta=None, is_fusion=True, executor=self)
+        fusion_bsym = BoundSymbol(
+            fusion_sym,
+            inputs,
+            {},
+            outputs,
+            region.bound_symbols,
+            _call_ctx={fusion_name: fusion_callable},
+        )
+
+        return fusion_bsym
+
+    def can_fuse(self, bsym: BoundSymbol):
+        curr_tracectx = get_tracectx()
+        assert hasattr(curr_tracectx, "clear_collection_names")
+
+        # Related to backward traces.
+        # No CollectionProxies in fusions! {
+        # Arguments of `clear_mutable_collection` should not be fused
+        if bsym.sym.id == "clear_mutable_collection":
+            curr_tracectx.clear_collection_names.add(bsym.args[0].name)
+            return False
+
+        # We let DEL to get fused, unless the deleted proxy is a CollectionProxy
+        # consumed by the `clear_mutable_collection` symbol
+        if bsym.sym.id == prims.PrimIDs.DEL and bsym.args[0].name in curr_tracectx.clear_collection_names:
+            return False
+        # }
+
+        do_not_fuse_sym_set = {
+            # Skip the very beginning and the very end of the trace
+            prims.PrimIDs.UNPACK_TRIVIAL,
+            prims.PrimIDs.UNPACK_KEY,
+            prims.PrimIDs.UNPACK_EMPTY_DICT,
+            prims.PrimIDs.UNPACK_SEQUENCE,
+            prims.PrimIDs.RETURN,
+            # Data-dependent ops
+            prims.PrimIDs.ITEM,
+        }
+
+        if bsym.sym.id in do_not_fuse_sym_set:
+            return False
+
+        return True
```

Comment:
Would we need to check for fixed sizes of the proxies in input and output or ist this handled?

Reply:
I wonder whether the future API is agreed upon? Does it mean that dynamic shape Tensors will contain Integer proxies in their meta-data?